src/molecular_moments.F

```diff
@@ -0,0 +1,278 @@
+               END DO
+               !
+               CALL cp_fm_release(mvector)
+               CALL cp_fm_release(omvector)
+               CALL cp_fm_release(momv)
+               DEALLOCATE (states)
+            END DO
+            DO i = 1, nm
+               CALL dbcsr_deallocate_matrix(moments(i)%matrix)
+            END DO
+            DEALLOCATE (moments)
+         END DO
+      END IF
+      !
+      DO ikind = 1, molkind ! loop over different molecules
+         nmol = SIZE(local_molecules%list(ikind)%array)
+         DO imol = 1, nmol ! all the molecules of the kind
+            imol_now = local_molecules%list(ikind)%array(imol) ! index in the global array
+            molecule_kind => molecule_set(imol_now)%molecule_kind
+            first_atom = molecule_set(imol_now)%first_atom
+            CALL get_molecule_kind(molecule_kind=molecule_kind, natom=natom)
+            ! Get reference point for this molecule
+            CALL get_reference_point(rcc, qs_env=qs_env, reference=use_mom_ref_com, &
+                                     ref_point=ref_point, ifirst=first_atom, &
+                                     ilast=first_atom+natom-1)
+            ! charge
+            DO iatom = 1, natom
+               i = first_atom+iatom-1
+               atomic_kind => particle_set(i)%atomic_kind
+               CALL get_atomic_kind(atomic_kind, kind_number=akind)
+               CALL get_qs_kind(qs_kind_set(akind), ghost=ghost, floating=floating)
+               IF (.NOT. ghost .AND. .NOT. floating) THEN
+                  CALL get_qs_kind(qs_kind_set(akind), core_charge=zeff)
+                  charge_set(imol_now) = charge_set(imol_now)+zeff
+               END IF
+            END DO
+            DO ispin = 1, nspins
+               IF (ASSOCIATED(molecule_set(imol_now)%lmi(ispin)%states)) THEN
+                  ns = SIZE(molecule_set(imol_now)%lmi(ispin)%states)
+                  charge_set(imol_now) = charge_set(imol_now)-zwfc*ns
+               END IF
+            ENDDO
+            !
+            IF (norder > 0) THEN
+               ! nuclear contribution
+               DO i = 1, nm
+                  lx = indco(1, i+1)
+                  ly = indco(2, i+1)
+                  lz = indco(3, i+1)
+                  DO iatom = 1, natom
+                     j = first_atom+iatom-1
+                     atomic_kind => particle_set(j)%atomic_kind
+                     CALL get_atomic_kind(atomic_kind, kind_number=akind)
+                     CALL get_qs_kind(qs_kind_set(akind), ghost=ghost, floating=floating)
+                     IF (.NOT. ghost .AND. .NOT. floating) THEN
+                        CALL get_qs_kind(qs_kind_set(akind), core_charge=zeff)
+                        ria = particle_set(j)%r-rcc
+                        ria = pbc(ria, cell)
+                        moment_set(i, imol_now) = moment_set(i, imol_now)+ &
+                                                  zeff*ria(1)**lx*ria(2)**ly*ria(3)**lz
```

Comment:
not sure whether it would be clearer, but a bit more concise and without need for the temporaries:

```fortran
moment_set(i, imol_now) = moment_set(i, imol_now) + zeff*PRODUCT(ria**indco(1:3, i+1))
```

Reply:
It's not essential for performance, and this way reflects the basic equation better.